feat(linter): add `read_to_arena_str` function

overlookmotel opened the pull request:
Alternative to #11816. Read a file from disk directly into an `Allocator`.

Unfortunately, involves a pretty egregious hack (see comments in code), but this appears to be the only way that stable Rust provides to read a file into uninitialized memory without UB. Bonkers!

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11823?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#11823** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11823?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11823?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#11822** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11822?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#11821** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11821?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11823).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented:
@camc314 I *think* this should work, but we'll need to check it doesn't regress performance before making a release that uses it.

There are issues in Rust repo about optimizations that make `fs::read` do less syscalls, and I'm not 100% sure whether this method will also get the benefit of those optimizations or not (or whether it makes any practical difference either way).

camc314 commented on crates/oxc_linter/src/utils/mod.rs:
```diff
@@ -131,3 +140,120 @@ pub fn read_to_string(path: &Path) -> io::Result<String> {
+    allocator: &Allocator,
+) -> io::Result<&[u8]> {
+    // Check file is not larger than `isize::MAX` bytes (the max size of an allocation)
+    if isize::try_from(size).is_err() {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "File is larger than `isize::MAX` bytes",
+        ));
+    }
+    #[expect(clippy::cast_possible_truncation)]
+    let mut size = size as usize;
```
nit: rewrite as `let Ok(size) = isize::try_from(size) else {`

overlookmotel commented on crates/oxc_linter/src/utils/mod.rs:
```diff
@@ -131,3 +140,120 @@ pub fn read_to_string(path: &Path) -> io::Result<String> {
+    path: &Path,
+    allocator: &'alloc Allocator,
+) -> io::Result<&'alloc str> {
+    let mut file = File::open(path)?;
```
Weird that clippy doesn't flag this while the function is `#[expect(unused)]`. Fixed.

camc314 commented:
```
$ hyperfine --warmup 3 -n "main" '../../Boshen/oxc/target/release/oxlint --silent' -n "read str into allocator" '../../Boshen/oxc/target/release/str-into-alloc --silent'  -i
Benchmark 1: main
  Time (mean ± σ):     319.6 ms ±   2.3 ms    [User: 3132.9 ms, System: 185.8 ms]
  Range (min … max):   315.6 ms … 322.7 ms    10 runs

  Warning: Ignoring non-zero exit code.

Benchmark 2: read str into allocator
  Time (mean ± σ):     318.7 ms ±   3.3 ms    [User: 3126.3 ms, System: 181.0 ms]
  Range (min … max):   315.4 ms … 326.6 ms    10 runs

  Warning: Ignoring non-zero exit code.

Summary
  read str into allocator ran
    1.00 ± 0.01 times faster than main
$ hyperfine --warmup 3 -n "main" '../../Boshen/oxc/target/release/oxlint --silent' -n "read str into allocator" '../../Boshen/oxc/target/release/str-into-alloc --silent'  -i
Benchmark 1: main
  Time (mean ± σ):     322.2 ms ±   5.6 ms    [User: 3128.5 ms, System: 185.6 ms]
  Range (min … max):   316.0 ms … 332.7 ms    10 runs

  Warning: Ignoring non-zero exit code.

Benchmark 2: read str into allocator
  Time (mean ± σ):     323.3 ms ±   8.8 ms    [User: 3126.2 ms, System: 180.1 ms]
  Range (min … max):   315.3 ms … 336.4 ms    10 runs

  Warning: Ignoring non-zero exit code.

Summary
  main ran
    1.00 ± 0.03 times faster than read str into allocator
```

perf appears to be identical (or within noise limits)

camc314 commented:
### Merge activity

* **Jun 20, 9:03 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
feat(linter): add `read_to_arena_str` function (#11823)

Alternative to #11816. Read a file from disk directly into an `Allocator`.

Unfortunately, involves a pretty egregious hack (see comments in code), but this appears to be the only way that stable Rust provides to read a file into uninitialized memory without UB. Bonkers!
